### PR TITLE
[merge jam] #49285 - Improve MakeStack file parse error handling

### PR DESCRIPTION
**Pull request by waynew**
### What does this PR do?

Ports #49285

### Previous Behavior

TODO

### New Behavior

TODO

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/pillar/makostack.py [hunk 1]
```diff
@@ -9,24 +9,27 @@
...
 - all these rendered files are then parsed as ``yaml``.
-- then all yaml dicts are merged in order, with support for the following.
-  merging strategies: ``merge-first``, ``merge-last``, ``remove``, and
-  ``overwrite``.
-- stack config files can be matched based on ``pillar``, ``grains``, or
+- then all yaml dicts are merged in order, with support for the following
+  merging strategies: ``merge-first``
+  ``merge-last``
+  ``remove``
+  ``overwrite``
```
Comment (twangboy) [inline salt/pillar/makostack.py:24]:
```
- then all yaml dicts are merged in order, with support for the following
  merging strategies:
  ``merge-first``
  ``merge-last``
  ``remove``
  ``overwrite``
  ```

File: salt/pillar/makostack.py [hunk 2]
```diff
@@ -50,214 +53,221 @@
...
+      - /path/to/production.cfg
 
 Select config files through grains|pillar|opts matching
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-You can also opt for a much more flexible configuration: MakoStack allows one to
+You can also use a much more flexible configuration:  MakoStack allows one to
 select the config files for the current minion based on matching values from
-either grains, or pillar, or opts objects.
+either grains, or pillar, or opts.
```
Comment (twangboy) [inline salt/pillar/makostack.py:65]: grains, pillar, or opts

File: salt/pillar/makostack.py [hunk 3]
```diff
@@ -50,214 +53,221 @@
...
 
 - ``core.yml``
 - ``osarchs/amd64.yml``
 - ``oscodenames/jessie.yml``
 - ``roles/db.yml``
 - ``minions/test-1-dev.yml``
 
-Before merging, every files above will be preprocessed as mako templates.
-The following variables are available in mako templating of ``yaml`` files:
+Before merging, every files above will be preprocessed as mako templates.  The
```
Comment (twangboy) [inline salt/pillar/makostack.py:182]: Either `all files` or `every file`

File: salt/pillar/makostack.py [hunk 4]
```diff
@@ -50,214 +53,221 @@
...
-Once every ``yaml`` files have been processed, the ``stack`` dict will contain
-your whole own pillar data, merged in order by MakoStack.
-So MakoStack ``ext_pillar`` returns the ``stack`` dict, the contents of which
-Salt takes care to merge in with all of the other pillars and finally return
-the whole pillar to the minion.
+other MakoStack values that have already been parsed and evaluated (from
+``yaml`` files earlier in the configuration) through the ``stack`` variable.
+
+Once a ``yaml`` file is processed by mako, we obtain a Python dict - let's call
+it ``yml_data``.  This ``yml_data`` dict is then merged into in the main
```
Comment (twangboy) [inline salt/pillar/makostack.py:202]: `merged into the main`

File: salt/pillar/makostack.py [hunk 5]
```diff
@@ -50,214 +53,221 @@
...
+``stack`` dict (which itself is the already merged MakoStack pillar data).,
```
Comment (twangboy) [inline salt/pillar/makostack.py:203]: `.,` at the end

File: salt/pillar/makostack.py [hunk 6]
```diff
@@ -50,214 +53,221 @@
...
+based on the declared ``merge-strategy``.  By default, MakoStack will deeply
+merge ``yml_data`` into ``stack`` (much like the ``recurse`` option for salt's
```
Comment (twangboy) [inline salt/pillar/makostack.py:205]: `Salt's`

File: salt/pillar/makostack.py [hunk 7]
```diff
@@ -50,214 +53,221 @@
...
+``pillar_source_merging_strategy``), but 3 other merging strategies (see next
+section) are also available, on a per-object basis, to give you full control
+over the rendered data.
+
+Once all ``yaml`` files have been processed, the ``stack`` dict will contain
+MakoStack's copmlete pillar data.  At this point the MakoStack ``ext_pillar``
```
Comment (twangboy) [inline salt/pillar/makostack.py:211]: complete

File: salt/pillar/makostack.py [hunk 8]
```diff
@@ -50,214 +53,221 @@
...
+strategies are:
+- ``merge-last`` (the default)
+- ``merge-first``
+- ``remove``
+- ``overwrite``
 
-Note that scalar values like strings, integers, booleans, etc. are always
-evaluated using the ``overwrite`` strategy (other strategies don't make sense
-in that case).
+Note that scalar values like strings, integers, booleans, etc. (`leaf nodes` in
```
Comment (twangboy) [inline salt/pillar/makostack.py:226]:
```
``leaf nodes``
```

File: salt/pillar/makostack.py [hunk 9]
```diff
@@ -338,7 +348,7 @@
 |                      |         uid: 1001     |                         |
 +----------------------+-----------------------+-------------------------+
 
-You can also select a custom merging strategy using a ``__`` object in a list:
+Similarly, list allow a custom merging strategy using a ``__`` item:
```
Comment (twangboy) [inline salt/pillar/makostack.py:351]: `lists allow` or `list allows`?

File: salt/pillar/makostack.py [hunk 10]
```diff
@@ -370,6 +380,38 @@
 |       - tom    |       - __: overwrite   |       - mat             |
 |       - root   |       - mat             |                         |
 +----------------+-------------------------+-------------------------+
+
+Tweaking MakoStack
+------------------
+
+Out of the box, MakoStack (following the ``stack`` module it was cribbed from),
+will more or less silently pass over template files it cannot load due to mako
+or yaml parsing errors.  This is convenient, but arguably WRONG, behaviour; but
```
Comment (twangboy) [inline salt/pillar/makostack.py:389]: behavior

File: tests/pytests/functional/pillar/test_makostack.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
+import pytest
+import salt.pillar
+
+
+@pytest.fixture()
+def makostack_basic_config(tmp_path):
+    mako_config = tmp_path / "mako" / "mako.cfg"
```
Comment (twangboy) [inline tests/pytests/functional/pillar/test_makostack.py:7]: Would you want to do a `with tmp_path` here so that it gets cleaned up?

**Review by twangboy — CHANGES_REQUESTED** [review]

File: tests/pytests/functional/pillar/test_makostack.py (hunk 11 above)
Comment (waynew) [reply tests/pytests/functional/pillar/test_makostack.py:7]: tmp_path is (or should be) automatically cleaned up by the fixture. If it's not then that's broken across the board :upside_down_face:

File: salt/pillar/makostack.py (hunk 2 above)
Comment (twangboy) [reply salt/pillar/makostack.py:66]: Still need to fix this

File: salt/pillar/makostack.py (hunk 4 above)
Comment (twangboy) [reply salt/pillar/makostack.py:203]: And this

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/pillar/makostack.py (hunk 2 above)
Comment (waynew) [reply salt/pillar/makostack.py:66]: OOooo. I was looking for "opts objects" lol.

**Review by twangboy — DISMISSED** [review]

File: salt/pillar/makostack.py [hunk 12]
```diff
@@ -35,12 +39,12 @@
...
 
 This is the simplest option, you just need to set the path to your single
-MakoStack config file like below:
+MakoStack config file as shown below:
 
 .. code:: yaml
 
     ext_pillar:
-      - makostack: /path/to/stack.cfg
+    - makostack: /path/to/stack.cfg
```
Comment (Ch3LL) [inline salt/pillar/makostack.py:47]: the ext_pillar documentation here https://docs.saltproject.io/en/latest/topics/development/modules/external_pillars.html  includes the spaces. I think we should keep things consistent

File: salt/pillar/makostack.py [hunk 13]
```diff
@@ -50,214 +54,221 @@
 .. code:: yaml
 
     ext_pillar:
-      - makostack:
-          - /path/to/stack1.cfg
-          - /path/to/stack2.cfg
+    - makostack:
```
Comment (Ch3LL) [inline salt/pillar/makostack.py:57]: same here

File: salt/pillar/makostack.py [hunk 14]
```diff
@@ -50,214 +54,221 @@
...
-          pillar:environment:
-            dev: /path/to/dev/stack.cfg
-            prod: /path/to/prod/stack.cfg
-          grains:custom:grain:
-            value:
-              - /path/to/stack1.cfg
-              - /path/to/stack2.cfg
-          opts:custom:opt:
-            value: /path/to/stack0.cfg
+    - makostack:
```
Comment (Ch3LL) [inline salt/pillar/makostack.py:74]: and here

File: salt/pillar/makostack.py [hunk 15]
```diff
@@ -50,214 +54,221 @@
...
 
 - ``core.yml``
 - ``osarchs/amd64.yml``
 - ``oscodenames/jessie.yml``
 - ``roles/db.yml``
 - ``minions/test-1-dev.yml``
 
-Before merging, every files above will be preprocessed as mako templates.
-The following variables are available in mako templating of ``yaml`` files:
+Before merging, ever file above will be preprocessed as mako templates.  The
```
Comment (Ch3LL) [inline salt/pillar/makostack.py:183]:
```suggestion
Before merging, every file above will be preprocessed as mako templates.  The
```

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Comment by waynew** [comment]
Good catch, looks like I missed all of those.

File: salt/pillar/makostack.py (hunk 15 above)
Comment (Ch3LL) [reply salt/pillar/makostack.py:183]: I'm still seeing this

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**